### ISSUE TITLE
[Feature]: Support for DTS paths replacement

### ISSUE THREAD

**Issue by chenjiahan** [web-infra-dev/rslib#409]
### What problem does this feature solve?

The `compilerOptions#paths` in tsconfig.json should work for DTS output.

Modern.js Module supports this, see https://github.com/web-infra-dev/modern.js/blob/main/packages/solutions/module-tools/src/utils/dts.ts#L116

- tsconfig.json:

```json
{
  "compilerOptions": {
    "paths": {
      "rslog": ["./compiled/rslog"]
    }
  },
  "include": ["src"]
}
```

- src/index.ts:

```ts
import logger from 'rslog';
```

- dist/index.d.ts:

```ts
import logger from '../compiled/rslog';
```

### What does the proposed API look like?

This should work out of the box.

**Comment by Timeless0911** [comment]
Sub task of #140

**Comment by noshower** [comment]
现在 0.1.5 版本解决了值导入的别名问题，还差类型导入的别名问题没实现，加油

**Comment by tutods** [comment]
Any update on this?

**Comment by Timeless0911** [comment]
> Any update on this?

Not yet, you can use [tsc-alias](https://www.npmjs.com/package/tsc-alias) to replace alias paths with relative paths after Rslib compilation.

**Comment by tutods** [comment]
> > Any update on this?
> 
> Not yet, you can use [tsc-alias](https://www.npmjs.com/package/tsc-alias) to replace alias paths with relative paths after Rslib compilation.

But when I try to use the `tsc-alias` we move all the `.d.ts` to `dist/src` while the other files are on `dist` root.

**Comment by Timeless0911** [comment]
> But when I try to use the `tsc-alias` we move all the `.d.ts` to `dist/src` while the other files are on `dist` root.

This is usually due to `tsconfig.json` wrongly configured, to remove `src` folder level, we can configure like below:

```json
{
  "compilerOptions": {
    "baseUrl": ".",
    "rootDir": "src",
  },
  "include": ["src"]
}
```

**Comment by lachieh** [comment]
@Timeless0911 is this something that is open for contribution? I have an POC plugin using `tsc-alias` that works (with the limited testing I've done)

**Comment by Timeless0911** [comment]
> [@Timeless0911](https://github.com/Timeless0911) is this something that is open for contribution? I have an POC plugin using `tsc-alias` that works (with the limited testing I've done)

Yeah, PR is welcome for this feature which should be a config named `redirect.dts.path`. 

TBH, I wanna to implement this feature by someway using AST formly. And `tsc-alias` is also a solution based on regular expression which may exist some edge case. 

I think you can implement this feature in a simple way first, and then we can iterate on it gradually day by day.

**Comment by lachieh** [comment]
Ok great, I'll take a look. Agreed on AST implementation. Is there a recommended place to start looking into that? I hit some issues with a similar contribution to `vite-plugin-dts`.

There are definitely some edge cases with simple regular expressions like having `paths: { "*": "./src/*" }` in the tsconfig which is valid. Without some extra checks, that means that `import React from "react"` would get changed to `import React from "./src/react"`.

I'll make sure to handle the same cases I covered for the vite plugin, but AST as well as some extra resolution might be required.

**Comment by fi3ework** [comment]
FWIW, es-module-lexer and cjs-modules-lexer are way cheaper and accurate way to find the importing statements comparing to regex and AST parsing.

**Comment by Timeless0911** [comment]
@lachieh Hi, we need to put this feature on the agenda this quarter. I think we can first implement a simple version to ensure the availability of basic scenes, and then gradually iterate and optimize it. Are you interested in contributing a PR to implement it and then we can complete this feature together.

**Comment by lachieh** [comment]
Thanks for the ping! Apologies for not getting to it, I would be able to look at it early next week if that works?

I like the look of the lexer packages, does an implementation using those sound like a good first iteration?

**Comment by fi3ework** [comment]
It suddenly occurred to me, is it possible to achieve this through a TypeScript plugin?

**Comment by Timeless0911** [comment]
I think both TypeScript plugin and lexer are good for me to implement the first iteration.

**Comment by Timeless0911** [comment]
@lachieh Hi, I have submitted a PR as the first iteration which use `@ast-grep/napi` since `es-module-lexer` does not support `export type`. We can enhance it in the future.

**Comment by lachieh** [comment]
Thank you, @Timeless0911! I was going to hand this back off due to lack of time. I'll give the branch a trial run.

**Comment by yf-yang** [comment]
@Timeless0911 Awesome work! Thank you for your efforts! When could rslib release a new version with this PR included?

**Comment by Timeless0911** [comment]
> When could rslib release a new version with this PR included?

@yf-yang After Rspack releases a new version this week, you can use `0.5.0-alpha.0` first.

**Comment by yf-yang** [comment]
Thank you for all of your efforts! That's really really really helpful ❤
